rfcs/005-manual-grading-mvp.md

```diff
@@ -0,0 +1,91 @@
+
+When developing manual grading, features should integrate with PraireLearn’s existing grading architecture--autograders should still run if possible, the grade function should still run, ect.
+
+### Sample Use Cases
+
+- Provide feedback for and help easily grade elements that are difficult to autograde (i.e. longform text responses, file uploads).
+- Trigger a manual grading workflow for a certain question part if an autograder fails.
+- Grant instructors the option to regrade parts of an autograded question for style points.
+- Allow easy, granular, and descriptive manual grading via CSV upload.
+- Retrigger grading workflows after updating a question’s grade function.
+
+## Goals and Objectives
+
+### Core Goals and Objectives
+
+#### **A way to mark questions for “manual grading”**
+
+- To create an extensible manual grading feature which can be applied to _all_ elements, including file upload and custom elements, PraireLearn needs a generic way to mark a question for manual grading.
+  - We can achieve this through a wrapping element or through a “manual=true” flag on elements. The latter encourages developing element inheritance, and implicitly attaching scores, argument values, ect. to any element with the “manual=true” flag.
+- In order to clearly communicate with students, we need a way to inform them that question grades can still change after the autograder grants a preliminary score.
+  - PrairieLearn has multiple question states already, which were initially tied into the "pending autograde"/"in autograde"/"graded" workflow. For consistency, we should look at appropriating this.
+- To take advantage of autograders, we should be able to mark a question as “finished grading” if an autograder passes without issue, and mark it as “unfinished grading” if tests fail.
+
+#### **A common regrade** **workflow**
+
+- We have multiple use cases for “regrading” or “manual grading” within PraireLearn, so we need a common workflow to do so.
+  - Many instructors wish to recalculate grades after adjusting the grade function, or making other alterations to the question.
+  - When uploading scores in the CSV re-grading interface, it is useful to provide scores for particular parts of questions, so instructors do not have to try to recalculate question totals themselves.
+  - With a manual grading interface, an instructor may want to update scores for a particular part of the question.
+- To avoid race conditions of total score calculations v partial score calculations, we should introduce a secondary “grade” function, breaking up the grading into a “pre” part to compute partial scores, and a “post” part to calculate a final score.
+
```

Comment:
Lines 45-48 can be taken out, as that is a remnant of an image taken from Slack.